### PR TITLE
flow: add support to define elephant flows

### PR DESCRIPTION
**Pull request by inashivb**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/5647

Known TODOs:
1. Define and implement what marking an elephant flow from a rule would look like.
2. Steps that come after a flow has been marked. Should it be deferred to FM? What shall be done? Should it be a generic solution or configurable, etc.

Looking for feedback on:
1. Generic flags seemed to be full so added this as a separate boolean in `Flow` struct. Is this a good decision?
2. Is a default of u32_max ok?
3. Should the setting be off by default in `suricata.yaml`?

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 ERROR: ASAN TEST FAIL in ASAN_TLPR1_suri QA test

ERROR: QA failed on ASAN_TLPR1_suri.
  

Pipeline 22168

**Comment by suricata-qa** [comment]
ERROR:
 ERROR: ASAN TEST FAIL in ASAN_TLPR1_suri QA test

ERROR: QA failed on ASAN_TLPR1_suri.
  

Pipeline 22169

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
  

Pipeline 22170

File: suricata.yaml.in [hunk 1]
```diff
@@ -1474,6 +1474,13 @@ flow:
   #managers: 1 # default to one flow manager
   #recyclers: 1 # default to one flow recycler thread
 
+# Mark a flow as an elephant flow if the rate of influx of bytes is equal to
+# or exceeds the given number. The rate is calculated by dividing the total
+# number of bytes a flow has seen in either direction divided by the time a
```
Comment (inashivb) [inline suricata.yaml.in:1479]: s/either/both

File: src/flow.h [hunk 2]
```diff
@@ -301,6 +301,8 @@ typedef struct FlowCnf_
     uint32_t timeout_new;
     uint32_t timeout_est;
 
+    uint32_t elephant_flow_rate;
```
Comment (catenacyber) [inline src/flow.h:304]: It hurts to increase this structure...

File: suricata.yaml.in (hunk 1 above)
Comment (catenacyber) [reply suricata.yaml.in:1479]: time in seconds ?

Comment (inashivb) [reply suricata.yaml.in:1479]: yes. That's important to mention. I missed it.

File: src/flow.h (hunk 2 above)
Comment (inashivb) [reply src/flow.h:304]: I see. What is your suggestion?

**Comment by inashivb** [comment]
> > Define and implement what marking an elephant flow from a rule would look like.
> 
> For me, an elephant flow must last long, and have many bytes, not just be fast...

I see. Thank you. Will think more about accommodating these conditionals.

File: src/flow.h (hunk 2 above)
Comment (catenacyber) [reply src/flow.h:304]: What is elephant flow rate ? Cannot we compute it on demand instead of storing it ?

Comment (catenacyber) [reply src/flow.h:304]: What is elephant flow rate ? Cannot we compute it on demand instead of storing it ?

Comment (inashivb) [reply src/flow.h:304]: Just checked the code.
Q: Why does it hurt to increase this struct? It does not seem to be referenced/passed around inside other crucial structs and only stores config options wrt flow.

Comment (catenacyber) [reply src/flow.h:304]: Sorry, `FlowCnf_`is ok to expand, it is the other change below

File: src/flow.h [hunk 3]
```diff
@@ -503,6 +505,7 @@ typedef struct Flow_
     uint64_t todstbytecnt;
     uint64_t tosrcbytecnt;
 
+    bool elephant;
```
Comment (catenacyber) [inline src/flow.h:508]: It hurts to increase this structure...

Comment (catenacyber) [reply src/flow.h:508]: Did you check all flags were used ?

Comment (inashivb) [reply src/flow.h:508]: Seemed like that. Will check again.

Comment (inashivb) [reply src/flow.h:508]: yes they're all used

Comment (inashivb) [reply src/flow.h:508]: Do you have suggestions on where to store this then?

Comment (catenacyber) [reply src/flow.h:508]: Why do we need to store it ?
Can we not have a function FlowIsElephant(Flow *f) ?

Comment (jlucovsky) [reply src/flow.h:508]: `pahole` shows 2 gaps -- 2 and 4 bytes respectively.

Comment (inashivb) [reply src/flow.h:508]: > Why do we need to store it ? Can we not have a function FlowIsElephant(Flow *f) ?

I thought about it but then wondered if adding more to calculations each time byte count was incremented would be too much and ended up storing it in a boolean. Do you think it is ok to do that?

Comment (inashivb) [reply src/flow.h:508]: > `pahole` shows 2 gaps -- 2 and 4 bytes respectively.

Thank you. Will keep in mind to check that if I don't find a better solution.

**Comment by inashivb** [comment]
This implementation is flawed.
Case pointed out by Victor that is not covered by the current detection: The flows starts out fine but has a surge of pkts in the middle so the average rate as per the `flow.age` is fine but the spike suggested it should be an elephant flow.

**Comment by catenacyber** [comment]
> The flows starts out fine but has a surge of pkts in the middle so the average rate as per the flow.age is fine but the spike suggested it should be an elephant flow.

🤔 I would not have said this is an elephant flow...
So, what is the definition of an elephant flow ?

**Comment by inashivb** [comment]
> 🤔 I would not have said this is an elephant flow... 

Why? Can the sudden spike in traffic not overwhelm Suricata?

> So, what is the definition of an elephant flow ?

There are 2 for now:
1. The rate of influx of bytes is very high.
2. As suggested by you earlier, a very long lasting flow w lots of bytes.

resulting in high CPU utilization and/or pkt drops.

**Comment by catenacyber** [comment]
> > 🤔 I would not have said this is an elephant flow...
> 
> Why? Can the sudden spike in traffic not overwhelm Suricata?
> 

Yes, it can.
So, I guess we want to detect these dangerous flows, which are not what I would call elephant flows...
I guess you want to measure and bound the total CPU time spent on a flow...

> > So, what is the definition of an elephant flow ?
> 
> There are 2 for now:
> 
> 1. The rate of influx of bytes is very high.
> 2. As suggested by you earlier, a very long lasting flow w lots of bytes.
> 
> resulting in high CPU utilization and/or pkt drops.

Looks like 2 cousins : elephant and hippopotamus flows 😆 

How would you define "rate of influx bytes" ?

**Comment by inashivb** [comment]
> How would you define "rate of influx bytes" ?

Allow user to set the rate in terms of `x` number of bytes in `y` number of seconds. Currently, this PR also asks for that but in an incovenient way: asking exactly the rate.
`rate of influx bytes = x/y` for every `y` interval.
So, Victor suggested to look at a sliding window approach.
Let's say the user says they want to call a flow w 10k bytes in 10 seconds an elephant flow.
So, I should check if 10k bytes were received when flow was 10s old then, I should check if 10k bytes were received in 2-11 seconds (so still an interval of 10s) and so on.
Lmk wdyt?

**Comment by catenacyber** [comment]
This still looks weird to me, so I guess I should better remove myself from the discussion...

Looks like flows should have a measurement of how expensive they are in terms of CPU for suricata, and this should be used by the emergency mode somehow...

**Comment by inashivb** [comment]
> This still looks weird to me, so I guess I should better remove myself from the discussion...
> 

No, please. It just means that I need to work on my explanation and/or understanding. Either I explain it well enough that it also makes sense to you or I retrace the logical steps and find where am I incorrect.

> Looks like flows should have a measurement of how expensive they are in terms of CPU for suricata, and this should be used by the emergency mode somehow...

Yes. In the article you shared w me about elephant flows, by industry standards, that part is done to remediate the elephant flows so kind of an after step post detection iff the user wants it. I'll present something better soon and maybe you can see if it makes sense to you then. :)
Thanks a lot for spending time on this! 🙇🏽‍♀️

**Comment by inashivb** [comment]
Needs some rework to be presentable.

**Pull request closed without merging**